persistent-postgresql/test/UpsertWhere.hs

```diff
@@ -0,0 +1,111 @@
+import PgInit
+
+share [mkPersist sqlSettings, mkMigrate "upsertWhereMigrate"] [persistLowerCase|
+  Item
+     name        Text sqltype=varchar(80)
+     description Text
+     price       Double Maybe
+     quantity    Int Maybe
+
+     Primary name
+     deriving Eq Show Ord
+
+|]
+
+specs :: Spec
+specs = describe "UpsertWhere" $ do
+  let item1 = Item "item1" "" (Just 3) Nothing
+      item2 = Item "item2" "hello world" Nothing (Just 2)
+      items = [item1, item2]
+  describe "upsertWhere" $ do
+    it "inserts appropriately" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      upsertWhere item1 [ItemDescription =. "i am item 1"] []
+      Just item <- get (ItemKey "item1")
+      item @== item1
+
+    it "performs only updates given if record already exists" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      let newDescription = "I am a new description"
+      insert_ item1
+      upsertWhere
+        (Item "item1" "i am inserted description" (Just 1) (Just 2))
+        [ItemDescription =. newDescription] 
+        []
+      Just item <- get (ItemKey "item1")
+      item @== item1 { itemDescription = newDescription }
+
+  describe "upsertManyWhere" $ do
+    it "inserts fresh records" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      let newItem = Item "item3" "fresh" Nothing Nothing
+      upsertManyWhere
+        (newItem : items)
+        [copyField ItemDescription]
+        []
+        []
+      dbItems <- map entityVal <$> selectList [] []
+      sort dbItems @== sort (newItem : items)
+    it "updates existing records" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      upsertManyWhere
+        items
+        []
+        [ItemQuantity +=. Just 1]
+        []
+    -- TODO: this test doesn't pass
+    it "only copies passing values" $ runConnAssert $ do
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      let newItems = map (\i -> i { itemQuantity = Just 0, itemPrice = fmap (*2) (itemPrice i) }) items
+          postUpdate = map (\i -> i { itemPrice = fmap (*2) (itemPrice i) }) items
+      upsertManyWhere
+        newItems
+        [ copyUnlessEq ItemQuantity (Just 0)
+        , copyField ItemPrice
+        ]
+        []
+        []
+      dbItems <- sort . fmap entityVal <$> selectList [] []
+      dbItems @== sort postUpdate
+    it "inserts without modifying existing records if no updates specified" $ runConnAssert $ do
+      let newItem = Item "item3" "hi friends!" Nothing Nothing
+      deleteWhere ([] :: [Filter Item])
+      insertMany_ items
+      upsertManyWhere
+        (newItem : items)
+        []
+        []
+        []
+      dbItems <- sort . fmap entityVal <$> selectList [] []
+      dbItems @== sort (newItem : items)
+    -- TODO: this test doesn't pass
+    it "inserts without modifying existing records if no updates specified and there's a filter" $
```

Comment:
similar to the test above, this test also fails with a format error:
```
uncaught exception: FormatError
       FormatError {fmtMessage = "13 single '?' characters, but 12 parameters", fmtQuery = "INSERT INTO \"item\"(\"name\", \"description\", \"price\", \"quantity\") VALUES (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?) ON CONFLICT (\"name\") DO UPDATE SET \"name\"=EXCLUDED.\"name\" WHERE (\"description\"=?)", fmtParams = ["item3","hi friends!","null","null","item1","","3.0","null","item2","hello world","null","2"]}
```

I imagine this issue is similar to the one before; is it obvious what I'm doing wrong here?  I spent all afternoon on Friday looking at this and couldn't grok what I was doing wrong.

Reply:
here's the generated SQL for this failure
```
INSERT INTO "item"("name", "description", "price", "quantity") VALUES (?, ?, ?, ?), (?, ?, ?, ?), (?, ?, ?, ?) ON CONFLICT ("name") DO UPDATE SET "name"=EXCLUDED."name" WHERE ("description"=?)
```

Reply:
and the `PersistValue`
```
[PersistText "item3",PersistText "hi friends!",PersistNull,PersistNull,PersistText "item1",PersistText "",PersistDouble 3.0,PersistNull,PersistText "item2",PersistText "hello world",PersistNull,PersistInt64 2]
```